File: app/src/main/java/ru/santaev/clipboardtranslator/ui/adapter/LanguageAdapter.java
package ru.santaev.clipboardtranslator.ui.adapter;


import android.content.res.Resources;
import android.databinding.DataBindingUtil;
import android.support.v4.content.res.ResourcesCompat;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.List;

import ru.santaev.clipboardtranslator.R;
import ru.santaev.clipboardtranslator.TranslatorApp;
import ru.santaev.clipboardtranslator.databinding.LanguageItemLayoutBinding;
import ru.santaev.clipboardtranslator.databinding.LanguageSectionItemLayoutBinding;
import ru.santaev.clipboardtranslator.db.entity.Language;

public class LanguageAdapter extends RecyclerView.Adapter<LanguageAdapter.ViewHolder>{

    private static final int VIEW_TYPE_LANGUAGE = 0;
    private static final int VIEW_TYPE_SECTION = 1;

    public interface OnItemClickedListener{
        void onClick(Language translation);
    }

    private List<Language> recentLanguages;
    private List<Language> languages;
    private List<Language> unsupportedLanguages;
    private OnItemClickedListener listener;
    private boolean hasHeaders = false;

    public LanguageAdapter(List<Language> recentLanguages, List<Language> languages,
                           List<Language> unsupportedLanguages) {
        this.recentLanguages = recentLanguages;
        this.languages = languages;
        this.unsupportedLanguages = unsupportedLanguages;
        hasHeaders = !recentLanguages.isEmpty();
    }

    public OnItemClickedListener getListener() {
        return listener;
    }

    public void setListener(OnItemClickedListener listener) {
        this.listener = listener;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        if (viewType == VIEW_TYPE_LANGUAGE) {
            LanguageItemLayoutBinding binding = DataBindingUtil.inflate(LayoutInflater.from(parent.getContext()),
                    R.layout.language_item_layout, parent, false);
            return new LanguageViewHolder(binding);
        } else {
            LanguageSectionItemLayoutBinding binding = DataBindingUtil.inflate(LayoutInflater.from(parent.getContext()),
                    R.layout.language_section_item_layout, parent, false);
            return new SectionViewHolder(binding);
        }

    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        int viewType = getItemViewType(position);

        if (viewType == VIEW_TYPE_LANGUAGE) {
            boolean isRecentLang = hasHeaders && position - 1 < recentLanguages.size();
            int langPosition = position - (hasHeaders ? recentLanguages.size() + 2 : 0);
            boolean isSupported = langPosition < languages.size();

            LanguageViewHolder viewHolder = (LanguageViewHolder) holder;
            Language language = isRecentLang
                    ? recentLanguages.get(position - 1)
                    : getLangByPos(langPosition);


            viewHolder.binding.text.setText(language.getName());

            int color = isSupported ? R.color.textColorGrayDark : R.color.textColorGrayLight;
            Resources resources = TranslatorApp.Companion.getAppContext().getResources();
            viewHolder.binding.text.setTextColor(ResourcesCompat.getColor(resources, color, null));

            viewHolder.binding.getRoot().setOnClickListener(v -> {
                Language langByPos = isRecentLang
                        ? recentLanguages.get(position - 1)
                        : getLangByPos(langPosition);
                if (listener != null && null != langByPos) {
                    listener.onClick(langByPos);
                }
            });
        } else {
            SectionViewHolder viewHolder = (SectionViewHolder) holder;
            viewHolder.binding.text.setText(0 == position
                    ? R.string.choose_lang_activity_subheader_recent
                    : R.string.choose_lang_activity_subheader_langs);
        }

    }

    @Override
    public int getItemViewType(int position) {
        if (hasHeaders) {
            if (position == 0) {
                return VIEW_TYPE_SECTION;
            } else if (position < recentLanguages.size() + 1) {
                return VIEW_TYPE_LANGUAGE;
            } else if (position == recentLanguages.size() + 1) {
                return VIEW_TYPE_SECTION;
            } else {
                return VIEW_TYPE_LANGUAGE;
            }
        } else {
            return VIEW_TYPE_LANGUAGE;
        }
    }

    @Override
    public int getItemCount() {
        return languages != null
                ? languages.size() + unsupportedLanguages.size() + recentLanguages.size()
                + (hasHeaders ? 2 : 0)
                : 0;
    }

    private Language getLangByPos(int position){
        return position < languages.size()
                ? languages.get(position)
                : unsupportedLanguages.get(position - languages.size());
    }


    class ViewHolder extends RecyclerView.ViewHolder{

        ViewHolder(View itemView) {
            super(itemView);
        }
    }

    private class LanguageViewHolder extends ViewHolder{

        LanguageItemLayoutBinding binding;

        LanguageViewHolder(LanguageItemLayoutBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }

    private class SectionViewHolder extends ViewHolder {

        LanguageSectionItemLayoutBinding binding;

        SectionViewHolder(LanguageSectionItemLayoutBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }
}

File: app/src/main/java/ru/santaev/clipboardtranslator/ui/adapter/LanguageAdapter.kt
package ru.santaev.clipboardtranslator.ui.adapter


import android.databinding.DataBindingUtil
import android.support.v4.content.res.ResourcesCompat
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.example.santaev.domain.dto.LanguageDto
import ru.santaev.clipboardtranslator.R
import ru.santaev.clipboardtranslator.TranslatorApp
import ru.santaev.clipboardtranslator.databinding.LanguageItemLayoutBinding
import ru.santaev.clipboardtranslator.databinding.LanguageSectionItemLayoutBinding

class LanguageAdapter(
        private val recentLanguages: List<LanguageDto>,
        private val languages: List<LanguageDto>?,
        private val unsupportedLanguages: List<LanguageDto>
) : RecyclerView.Adapter<LanguageAdapter.ViewHolder>() {

    var listener: ((LanguageDto) -> Unit)? = null
    private var hasHeaders = false

    init {
        hasHeaders = !recentLanguages.isEmpty()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        if (viewType == VIEW_TYPE_LANGUAGE) {
            val binding = DataBindingUtil.inflate<LanguageItemLayoutBinding>(LayoutInflater.from(parent.context),
                    R.layout.language_item_layout, parent, false)
            return LanguageViewHolder(binding)
        } else {
            val binding = DataBindingUtil.inflate<LanguageSectionItemLayoutBinding>(LayoutInflater.from(parent.context),
                    R.layout.language_section_item_layout, parent, false)
            return SectionViewHolder(binding)
        }

    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val viewType = getItemViewType(position)

        if (viewType == VIEW_TYPE_LANGUAGE) {
            val isRecentLang = hasHeaders && position - 1 < recentLanguages.size
            val langPosition = position - if (hasHeaders) recentLanguages.size + 2 else 0
            val isSupported = langPosition < languages!!.size

            val viewHolder = holder as LanguageViewHolder
            val language = if (isRecentLang)
                recentLanguages[position - 1]
            else
                getLangByPos(langPosition)


            viewHolder.binding.text.text = language.name

            val color = if (isSupported) R.color.textColorGrayDark else R.color.textColorGrayLight
            val resources = TranslatorApp.appContext.resources
            viewHolder.binding.text.setTextColor(ResourcesCompat.getColor(resources, color, null))

            viewHolder.binding.root.setOnClickListener { v ->
                val langByPos = if (isRecentLang)
                    recentLanguages[position - 1]
                else
                    getLangByPos(langPosition)
                listener?.invoke(langByPos)
            }
        } else {
            val viewHolder = holder as SectionViewHolder
            viewHolder.binding.text.setText(if (0 == position)
                R.string.choose_lang_activity_subheader_recent
            else
                R.string.choose_lang_activity_subheader_langs)
        }

    }

    override fun getItemViewType(position: Int): Int {
        return if (hasHeaders) {
            when {
                position == 0 -> VIEW_TYPE_SECTION
                position < recentLanguages.size + 1 -> VIEW_TYPE_LANGUAGE
                position == recentLanguages.size + 1 -> VIEW_TYPE_SECTION
                else -> VIEW_TYPE_LANGUAGE
            }
        } else {
            VIEW_TYPE_LANGUAGE
        }
    }

    override fun getItemCount(): Int {
        return if (languages != null) {
            languages.size + unsupportedLanguages.size + recentLanguages.size + if (hasHeaders) 2 else 0
        } else {
            0
        }
    }

    private fun getLangByPos(position: Int): LanguageDto {
        return if (position < languages!!.size)
            languages[position]
        else
            unsupportedLanguages[position - languages.size]
    }


    open inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)

    private inner class LanguageViewHolder
    internal constructor(internal var binding: LanguageItemLayoutBinding) : ViewHolder(binding.root)

    private inner class SectionViewHolder
    internal constructor(internal var binding: LanguageSectionItemLayoutBinding) : ViewHolder(binding.root)

    companion object {

        private val VIEW_TYPE_LANGUAGE = 0
        private val VIEW_TYPE_SECTION = 1
    }
}
